openocd.board.cmake: Fix flash load address

qianfan-Zhao opened the pull request:
'make flash' always erase the flash banks from sector 0, and flashing
the Zephyr.bin file to base address of flash even if there has a
bootloader placed on the first sector.

Users maybe want placed Zephyr to another location instead of sector
0 by adding 'zephyr,code-partition' in dts. But the openocd scripts
doesn't deal with that currently.

Fix this issue, flash Zephyr.bin to address:
CONFIG_FLASH_BASE_ADDRESS + CONFIG_FLASH_LOAD_OFFSET.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
openocd.board.cmake: Fix flash load address

'make flash' always erase the flash banks from sector 0, and flashing
the Zephyr.bin file to base address of flash even if there has a
bootloader placed on the first sector.

Users maybe want placed Zephyr to another location instead of sector
0 by adding 'zephyr,code-partition' in dts. But the openocd scripts
doesn't deal with that currently.

Fix this issue, flash zephyr.elf instead of zephyr.bin, elf file
contain load address and openocd scripts can deal with it.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

SebastianBoe commented:
Note that this problem is specific to flashing bin files. Using hex files does not have this problem as it has addressing information.

I don't suppose it would be possible to solve this by using .hex instead of .bin?

.hex has other benefits as well, such as permitting holes so that one doesn't have to write more than necessary.

mbolivar commented:
> I don't suppose it would be possible to solve this by using .hex instead of .bin?

Good question. It is indeed. Both options (`load_image` and `flash write_image`) can operate on hex files. I think it would make sense to move there as well.

aurel32 commented:
openocd also supports flashing elf files, so that's another option which doesn't require a conversion.

I am not sure what is the best implementation, but the idea makes sense. It's also something useful when modifying the boot vector of a SoC. For example on the STM32F7 the flash has small sectors at the beginning, followed by bigger sectors. It therefore makes sense to place the storage partition before the code partition and to modify the boot vector accordingly. Modifying the partitions in an overlay DTS file almost works, just the flashing part has to be modified like in your patch.

qianfan-Zhao commented:
Since openocd support flash elf file directly, I prefer replace the scripts to using zephyr.elf file not zephyr.bin file.

qianfan-Zhao commented:
Hi:
I had changed this PR, flash Zephyr.elf instead of Zephyr.bin file. So I think the variable **OPENOCD_ADDRESS** are unused now, and deleted it.

I searched in openocd's documents, here is the introduction about command 'verify_iamge':
```
verify_image filename address [bin|ihex|elf]
[Command]
Verify filename against target memory starting at address. The file format may
optionally be specified (bin, ihex, or elf) This will first attempt a comparison using
a CRC checksum, if this fails it will try a binary compare.
```

Does the command really need param **address**? Now **make flash** worked fine.

SebastianBoe commented:
> Does the param 'address' in verify_image is really need?

Perhaps it's a bug in the docs, address should have been [address].

aurel32 reviewed (approved):
Thanks for this new version. I have tested it on Atmel SAM E70 Xplained and on a STM32F723E Disco board, the later with both the default reset vector and a modified reset vector with the associated DTS change. So it looks all good to me.
